File: src/tunif.c

//#define _BSD_SOURCE

#include <arpa/inet.h>
#include <fcntl.h>
#include <linux/if_tun.h>
//#include <linux/in.h>
#include <net/if.h>
//#include <socket.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>

// https://www.kernel.org/doc/Documentation/networking/tuntap.txt
void set_interface_name(int if_fd, const char *ifname)
{
    struct ifreq ifr = {};

    memset(&ifr, 0, sizeof(ifr));
    strncpy(ifr.ifr_name, ifname, IFNAMSIZ);
    ifr.ifr_flags = IFF_TUN | IFF_NO_PI;

    int err = ioctl(if_fd, TUNSETIFF, (void *) &ifr);
    if (err < 0)
    {
        perror("ERROR TUNSETIFF");
        exit(EXIT_FAILURE);
    }
}


void set_interface_address(int if_fd, const char *ifname, const char *addr, int netmask)
{
    (void)if_fd;
    struct ifreq ifr;

    struct in_addr ipv4;
    const struct in_addr nmask = {
        htonl((-1) ^ ((1<<(32-netmask))-1))
    };
    struct sockaddr_in ipv4_addr;

    if (netmask <= 0 || netmask >= 32)
    {
        fprintf(stderr, "Invalid netmask %d\n", netmask);
        exit(EXIT_SUCCESS);
    }

    memset(&ipv4, 0, sizeof(ipv4));
    if (inet_pton(AF_INET, addr, (void *)&ipv4.s_addr) != 1)
    {
        perror("Failed to parse address with inet_pton");
        exit(EXIT_FAILURE);
    }

    int udp_socket = socket(AF_INET, SOCK_DGRAM, 0);
    if (udp_socket < 0)
    {
        perror("Failed to create udp_socket");
        exit(EXIT_FAILURE);
    }

    memset(&ifr, 0, sizeof(ifr));
    strncpy(ifr.ifr_name, ifname, IFNAMSIZ);
    // set interface address
    memset(&ipv4_addr, 0, sizeof(ipv4_addr));
    ipv4_addr.sin_family = AF_INET;
    ipv4_addr.sin_addr = ipv4;
    memcpy(&ifr.ifr_addr, &ipv4_addr, sizeof(ipv4_addr));
    if (ioctl(udp_socket, SIOCSIFADDR, &ifr) < 0)
    {
        perror("Failed to set netaddr");
        if (close(udp_socket))
        {
            perror("DOUBLE FAULT close(udp_socket)");
        }
        exit(EXIT_FAILURE);
    }
    // set interface netmask
    memset(&ifr, 0, sizeof(ifr));
    strncpy(ifr.ifr_name, ifname, IFNAMSIZ);
    memset(&ipv4_addr, 0, sizeof(ipv4_addr));
    ipv4_addr.sin_family = AF_INET;
    ipv4_addr.sin_addr = nmask;
    memcpy(&ifr.ifr_addr, &ipv4_addr, sizeof(ipv4_addr));
    if (ioctl(udp_socket, SIOCSIFNETMASK, &ifr) < 0)
    {
        perror("Failed to set netmask");
        if (close(udp_socket))
        {
            perror("DOUBLE FAULT close(udp_socket)");
        }
        exit(EXIT_FAILURE);
    }

    if (close(udp_socket) < 0)
    {
        perror("Failed to close(udp_socket)");
        exit(EXIT_FAILURE);
    }
}

void set_interface_up(int if_fd, const char *ifname)
{
    (void)if_fd;
    // https://stackoverflow.com/questions/11679514/what-is-the-difference-between-iff-up-and-iff-running
    // difference IFF_UP vs IFF_RUNNING
    //  IFF_UP      =>  bring the interface up.
    //  IFF_RUNNING =>  check the interface status.
    struct ifreq ifr;
    memset(&ifr, 0, sizeof(ifr));
    strncpy(ifr.ifr_name, ifname, IFNAMSIZ);

    int udp_socket = socket(AF_INET, SOCK_DGRAM, 0);
    if (udp_socket < 0)
    {
        perror("Failed to create udp_socket");
        exit(EXIT_FAILURE);
    }

    // read interface flags
    if (ioctl(udp_socket, SIOCGIFFLAGS, &ifr) < 0)
    {
        perror("Failed to SIOCGIFFLAGS");
        if (close(udp_socket))
        {
            perror("DOUBLE FAULT close(udp_socket)");
        }
        exit(EXIT_FAILURE);
    }
    // set flags
    ifr.ifr_flags |= IFF_UP;
    if (ioctl(udp_socket, SIOCSIFFLAGS, &ifr) < 0)
    {
        perror("Failed to SIOCSIFFLAGS");
        if (close(udp_socket))
        {
            perror("DOUBLE FAULT close(udp_socket)");
        }
        exit(EXIT_FAILURE);
    }

    if (close(udp_socket) < 0)
    {
        perror("Failed to close(udp_socket)");
        exit(EXIT_FAILURE);
    }
}

void set_interface_down(int if_fd, const char *ifname)
{
    (void)if_fd;
    // https://stackoverflow.com/questions/11679514/what-is-the-difference-between-iff-up-and-iff-running
    // difference IFF_UP vs IFF_RUNNING
    //  IFF_UP      =>  bring the interface up.
    //  IFF_RUNNING =>  check the interface status.
    struct ifreq ifr;
    memset(&ifr, 0, sizeof(ifr));
    strncpy(ifr.ifr_name, ifname, IFNAMSIZ);

    int udp_socket = socket(AF_INET, SOCK_DGRAM, 0);
    if (udp_socket < 0)
    {
        perror("Failed to create udp_socket");
        exit(EXIT_FAILURE);
    }

    // read interface flags
    if (ioctl(udp_socket, SIOCGIFFLAGS, &ifr) < 0)
    {
        perror("Failed to SIOCGIFFLAGS");
        if (close(udp_socket))
        {
            perror("DOUBLE FAULT close(udp_socket)");
        }
        exit(EXIT_FAILURE);
    }
    // set flags
    ifr.ifr_flags &= ~IFF_UP;
    if (ioctl(udp_socket, SIOCSIFFLAGS, &ifr) < 0)
    {
        perror("Failed to SIOCSIFFLAGS");
        if (close(udp_socket))
        {
            perror("DOUBLE FAULT close(udp_socket)");
        }
        exit(EXIT_FAILURE);
    }

    if (close(udp_socket) < 0)
    {
        perror("Failed to close(udp_socket)");
        exit(EXIT_FAILURE);
    }
}

File: src/tunif.rs
use std::ffi::{CStr, CString};
use std::fs::File;
use std::net::{Ipv4Addr, SocketAddrV4};
use std::os::fd::{AsFd, AsRawFd};

use anyhow::{Result, bail};
use socket2::SockAddr;

#[inline(always)]
fn get_netmask(n: u8) -> u32 {
    assert!(n <= 32);
    0xFFFFFFFFu32.wrapping_shl(32 - n as u32)
}

struct Socket(i32);

impl Socket {
    fn new() -> Result<Self> {
        let socket = unsafe { libc::socket(libc::AF_INET, libc::SOCK_DGRAM, 0) };
        if socket < 0 {
            bail!("Error creating socket")
        }
        Ok(Socket(socket))
    }

    fn ioctl(&self, request: u64, mut ifr: libc::ifreq) -> Result<()> {
        let tmp: *mut libc::ifreq = &raw mut ifr;
        let res = unsafe { libc::ioctl(self.0, request, tmp) };
        if res < 0 {
            let err = std::io::Error::last_os_error();
            bail!("Error setting interface address: {}", err);
        }
        Ok(())
    }
}

impl Drop for Socket {
    fn drop(&mut self) {
        unsafe {
            libc::close(self.0);
        }
    }
}

unsafe fn ifr_create(ifname: &CStr, flags_fn: impl FnOnce(&mut libc::ifreq)) -> libc::ifreq {
    let mut ifr: libc::ifreq = unsafe { core::mem::zeroed() };
    for (i, &c) in ifname.to_bytes().iter().enumerate() {
        ifr.ifr_name[i] = c as i8;
    }
    ifr.ifr_name.last_mut().map(|c| *c = 0).unwrap();
    flags_fn(&mut ifr);
    ifr
}

fn set_interface_name(file: &File, ifname: &CStr) -> Result<()> {
    let ifr = unsafe {
        ifr_create(ifname, |ifr| {
            ifr.ifr_ifru.ifru_flags = (libc::IFF_TUN | libc::IFF_NO_PI) as i16;
        })
    };
    let res = unsafe { libc::ioctl(file.as_raw_fd(), libc::TUNSETIFF, &ifr) };
    if res < 0 {
        let err = std::io::Error::last_os_error();
        bail!("Error setting interface name: {}", err);
    }
    Ok(())
}

fn set_interface_address(socket: &Socket, ifname: &CStr, addr: &Ipv4Addr) -> Result<()> {
    let sockaddr: SockAddr = SocketAddrV4::new(*addr, 0).into();
    let tmp = unsafe { core::ptr::read(sockaddr.as_ptr()) };
    let ifr = unsafe { ifr_create(ifname, |ifr| ifr.ifr_ifru.ifru_addr = tmp) };
    socket.ioctl(libc::SIOCSIFADDR, ifr)?;
    Ok(())
}

fn set_subnet_mask(socket: &Socket, ifname: &CStr, netmask: u8) -> Result<()> {
    let mask = get_netmask(netmask).to_be_bytes();
    let ipv4addr = Ipv4Addr::new(mask[0], mask[1], mask[2], mask[3]);
    let sockaddr: SockAddr = SocketAddrV4::new(ipv4addr, 0).into();
    let tmp = unsafe { core::ptr::read(sockaddr.as_ptr()) };
    let ifr = unsafe { ifr_create(ifname, |ifr| ifr.ifr_ifru.ifru_netmask = tmp) };
    socket.ioctl(libc::SIOCSIFNETMASK, ifr)?;
    Ok(())
}

fn set_interface_up(socket: &Socket, ifname: &CStr) -> Result<()> {
    let ifr = unsafe {
        ifr_create(ifname, |ifr| {
            ifr.ifr_ifru.ifru_flags |= libc::IFF_UP as i16;
        })
    };
    socket.ioctl(libc::SIOCSIFFLAGS, ifr)?;
    Ok(())
}

fn set_interface_down(socket: &Socket, ifname: &CStr) -> Result<()> {
    let ifr = unsafe {
        ifr_create(ifname, |ifr| {
            ifr.ifr_ifru.ifru_flags &= !(libc::IFF_UP as i16);
        })
    };
    socket.ioctl(libc::SIOCSIFFLAGS, ifr)?;
    Ok(())
}

pub struct Iface {
    fd: File,
    name: CString,
    ip: Ipv4Addr,
    netmask: u8,
}

impl Iface {
    pub fn new(n: &str, ip: Ipv4Addr, netmask: u8) -> Result<Self> {
        if netmask > 32 {
            bail!("Netmask should be less than 32")
        }
        if n.len() > 16 {
            bail!("Interface name too long")
        }
        let fd = std::fs::OpenOptions::new()
            .read(true)
            .write(true)
            .open("/dev/net/tun")?;

        let name = CString::new(n)?;
        set_interface_name(&fd, &name)?;
        let socket = Socket::new()?;
        set_interface_address(&socket, &name, &ip)?;
        set_subnet_mask(&socket, &name, netmask)?;
        set_interface_up(&socket, &name)?;
        Ok(Iface {
            fd,
            name,
            ip,
            netmask,
        })
    }
}
impl AsRef<File> for Iface {
    fn as_ref(&self) -> &File {
        &self.fd
    }
}

impl AsFd for Iface {
    fn as_fd(&self) -> std::os::unix::prelude::BorrowedFd<'_> {
        self.fd.as_fd()
    }
}

impl Drop for Iface {
    fn drop(&mut self) {
        let f = || -> Result<()> { set_interface_down(&Socket::new()?, &self.name) };
        f().unwrap();
    }
}
